src/JsonRpcEngine.ts

```diff
@@ -296,18 +361,43 @@ export class JsonRpcEngine extends SafeEventEmitter {
         `Must specify a string method. Received: ${typeof callerReq.method}`,
         { request: callerReq },
       );
-      return callback(error, { id: callerReq.id, jsonrpc: '2.0', error });
+
+      if (this._notificationHandler && !isJsonRpcRequest(callerReq)) {
+        // Do not reply to notifications, even malformed ones.
+        return callback(null);
+      }
+
+      return callback(error, {
+        id: (callerReq as JsonRpcRequest<unknown>).id ?? null,
```

Comment:
Is there a way to get TypeScript to discriminate what `callerReq` is at this location without using `as`? Or are we typecasting for the same reason as given at line 393?

Reply:
No, it's because `callerReq` can be a notification, and it's easiest to try to access the `id` property even if it doesn't exist. I added a comment in 70327b1.

Reply:
Makes sense, thanks!